proposals/0000-import-shadowing.rst

```diff
@@ -0,0 +1,430 @@
+in the definition of ``goFishing``:
+
+::
+
+ Ambiguous occurrence `catch'
+ It could refer to
+    either `Control.Exception.catch',
+           imported from `Control.Exception' at Example1.hs:1:1-24
+           (and originally defined in `GHC.IO')
+        or `Example1.catch', defined at Example1.hs:7:1
+
+Our proposal is an alternative name resolution policy exposed as a
+language extension tentatively named ``ImportShadowing``. It allows
+definitions from the current module's top-level scope to shadow names
+from imported modules. So in this example, the occurrence of ``catch``
+in the definition of ``goFishing`` resolves to the definition in
+``Example1`` even when ``catch`` is not explicitly hidden from the
+import list of ``Control.Exception``.
+
+We believe this is an improvement to the status quo because it is a
+natural extension of the idea behind the shadowing policy of local
+binders: that shadowing expresses the intention by the user to only
+care about the names that are defined "more nearby". A cursory search
+on `Stack Overflow <https://stackoverflow.com/>`_ finds lots of
+Haskell users who implicitly expected imports to be shadowed by
+top-level definitions:
+
+* https://stackoverflow.com/q/73788349/477476
+* https://stackoverflow.com/q/7761238/477476
+* https://stackoverflow.com/q/56047335/477476
+* https://stackoverflow.com/q/40964909/477476
+* https://stackoverflow.com/q/40314142/477476
+* https://stackoverflow.com/q/67246392/477476
+
+A second-order effect of the proposed extension is that it can lead to
+preemptive forward compatibility. Adding a new export to ``Prelude``
+can lead to breakage just by virtue of existing code defining and
+using top-level definitions with the same name. With
+``ImportShadowing``, the existing intra-module references keep their
+meaning and there is no migration needed to accomodate the new
+``Prelude`` names.
+
+Proposed Change Specification
+-----------------------------
+
+When ``ImportShadowing`` is enabled, the following changes take place:
+
+Resolution of references in module body
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Consider an occurrence of an unqualified name ``x``, not bound locally
+(by ``let``, lambda, a ``case`` alternative, etc). There are two
+possible sources of resolving it:
+
+(A) If there is a top-level binding of ``x`` then the occurrence is
+    resolved to that binding.
+
+(B) If the import declarations bring into scope a unique entity with
+    unqualified name ``x``, the occurrence is resolved to that entity.
+
+Consider an occurrence of a qualified name ``M.x``:
+
+(A) If the module is called ``M`` and there is a top-level binding of
+    ``x``, the occurrence is resolved to that binding
+
+(B) If the import declarations bring into scope a unique entity with
+    qualified name ``M.x``, the occurrence is resolved to that
+    entity.
+
+In both cases, Haskell 2010 regards cases (A) and (B) on equal
```

Comment:
Please link to [Haskell 2010 report, Section 5.5.2](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html#x11-1090005.5.2).